r/R/dplyr-datetime-helpers.R

```diff
@@ -201,19 +213,130 @@ build_formats <- function(orders) {
+    )
+  }
+
+  # for `yq` we need to transform the quarter into the start month (lubridate
+  # behaviour) and then add 01 to parse to the first day of the quarter
+  augmented_x_yq <- NULL
+  if (any(orders == "yq")) {
+    # extract everything that comes after the `-` separator, i.e. the quarter
+    # (e.g. 4 from 2022-4)
+    quarter_x <- call_binding("gsub", "^.*?-", "", processed_x)
+    # we should probably error if quarter is not in 1:4
+    # extract everything that comes before the `-`, i.e. the year (e.g. 2002
+    # in 2002-4)
+    year_x <- call_binding("gsub", "-.*$", "", processed_x)
+    quarter_x <- quarter_x$cast(int32())
+    month_x <- (quarter_x - 1) * 3 + 1
+    augmented_x_yq <- call_binding("paste0", year_x, "-", month_x, "-01")
+  }
+
+  list(
+    "augmented_x_ym" = augmented_x_ym,
+    "augmented_x_yq" = augmented_x_yq,
+    "processed_x" = processed_x
+  )
+}
+
+attempt_parsing <- function(x,
+                            orders,
+                            formats = NULL) {
+  if (is.null(formats)) {
+    # this is the situation in which orders were passed with `exact = TRUE`
+    # no data processing takes place
+    # we don't derive formats as the orders are assumed to be formats
+    parse_attempt_expressions <- build_strptime_exps(x, orders)
+    return(parse_attempt_expressions)
+  }
+
+  processed_data <- process_data_for_parsing(x, orders)
+
+  processed_x <- processed_data[["processed_x"]]
+  augmented_x_ym <- processed_data[["augmented_x_ym"]]
+  augmented_x_yq <- processed_data[["augmented_x_yq"]]
+
+  # build a list of expressions for each format
+  parse_attempt_exp_processed_x <- build_strptime_exps(processed_x, formats)
+
+  # build separate expression lists of parsing attempts for the orders that
+  # need an augmented `x`
+  # list for attempts when orders %in% c("ym", "my")
+  parse_attempt_exp_augmented_x_ym <- list()
+
+  if (!is.null(augmented_x_ym)) {
+    parse_attempt_exp_augmented_x_ym <- build_strptime_exps(augmented_x_ym, formats)
+  }
+
+  # list for attempts when orders %in% c("yq")
+  parse_attempt_exp_augmented_x_yq <- list()
+  if (!is.null(augmented_x_yq)) {
+    parse_attempt_exp_augmented_x_yq <- build_strptime_exps(augmented_x_yq, formats)
+  }
+
+  # combine all attempts expressions in prep for coalesce
+  # if the users passes only a short order (`ym`, `my` or `yq`) then only use
+  # the corresponding augmented_x
+  if (all(orders == "ym") || all(orders == "my")) {
+    parse_attempt_expressions <- parse_attempt_exp_augmented_x_ym
+  } else if (all(orders == "yq")) {
+    parse_attempt_expressions <- parse_attempt_exp_augmented_x_yq
+  } else {
+    parse_attempt_expressions <- c(
+      # if we have an augmented x give preference to the corresponding
+      # parsing attempts
+      parse_attempt_exp_augmented_x_ym,
+      parse_attempt_exp_augmented_x_yq,
+      parse_attempt_exp_processed_x
+    )
+  }
+  parse_attempt_expressions
+}
```

Comment:
This is a pretty hairy chunk that's pretty hard to follow. We should make it a bit easier on our future selves to follow and maintain it by doing some combination of (or all) of the following:

* Adding slightly more detailed comments, especially explaining what general shape of things might be in the arguments — it's hard to know now what things are lists of expressions versus lists of characters like `"Ym"` versus  lists of characters like `"%Y-%m"`
* Refactor this code to work on more directly, instead of having sets of functions that produce lists that you need to subset out of, would it be possible to have functions that take as input (vectors of) single elements and operate on them to go from `"Ym"` to the expression, expanding as necessary.

Reply:
💯 You should have seen `parse_date_time` before the refactoring. I'll try to refactor even more.

Reply:
Done.